File: spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        t=0
        l=0
        b=len(matrix)-1
        r=len(matrix[0])-1
        while t<=b and l<=r:
            for i in range(l,r+1):
                ans.append(matrix[t][i])
            t+=1
            for i in range(t,b+1):
                ans.append(matrix[i][r])
            r-=1
            if t<=b:
                for i in range(r,l-1,-1):
                    ans.append(matrix[b][i])
                b-=1
            if l<=r:    
                for i in range(b,t-1,-1):
                    ans.append(matrix[i][l])
                l+=1
        return ans

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int n = matrix.length;
        int m = matrix[0].length;

        int left = 0, right = m - 1, top = 0, bottom = n - 1;

        while (left <= right && top <= bottom) {

            // moving left to right
            for (int i = left; i <= right; i++)
                ans.add(matrix[top][i]);
            top++;

            // moving top to bottom
            for (int i = top; i <= bottom; i++)
                ans.add(matrix[i][right]);
            right--;

            if (top <= bottom) {
                // moving right to left
                for (int i = right; i >= left; i--)
                    ans.add(matrix[bottom][i]);
                bottom--;
            }

            if(left<=right){
            // moving bottom to top
            for (int i = bottom; i >= top; i--)
                ans.add(matrix[i][left]);
            left++;
            }
        }
        return ans;
    }
}